chain/thorchain/thorchain.go

```diff
@@ -0,0 +1,1252 @@
+package thorchain
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	"github.com/cosmos/cosmos-sdk/crypto/hd"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	"github.com/cosmos/cosmos-sdk/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types" // nolint:staticcheck
+	chanTypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	dockertypes "github.com/docker/docker/api/types"
+	volumetypes "github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/client"
+	"github.com/strangelove-ventures/interchaintest/v8/blockdb"
+	"github.com/strangelove-ventures/interchaintest/v8/dockerutil"
+	"github.com/strangelove-ventures/interchaintest/v8/ibc"
+	"github.com/strangelove-ventures/interchaintest/v8/testutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Thorchain struct {
+	testName      string
+	cfg           ibc.ChainConfig
+	NumValidators int
+	numFullNodes  int
+	Validators    ChainNodes
+	FullNodes     ChainNodes
+
+	// preStartNodes is able to mutate the node containers before
+	// they are all started
+	preStartNodes func(*Thorchain)
+
+	// Additional processes that need to be run on a per-chain basis.
+	Sidecars SidecarProcesses
+
+	cdc      *codec.ProtoCodec
+	log      *zap.Logger
+	keyring  keyring.Keyring
+	findTxMu sync.Mutex
+}
+
+func NewThorchainHeighlinerChainConfig(
+	name string,
+	binary string,
+	bech32Prefix string,
+	denom string,
+	gasPrices string,
+	gasAdjustment float64,
+	trustingPeriod string,
+	noHostMount bool,
+) ibc.ChainConfig {
+	return ibc.ChainConfig{
+		Type:           "thorchain",
+		Name:           name,
+		Bech32Prefix:   bech32Prefix,
+		Denom:          denom,
+		GasPrices:      gasPrices,
+		GasAdjustment:  gasAdjustment,
+		TrustingPeriod: trustingPeriod,
+		NoHostMount:    noHostMount,
+		Images: []ibc.DockerImage{
+			{
+				Repository: "ghcr.io/strangelove-ventures/heighliner/thorchain",
+				UidGid:     dockerutil.GetHeighlinerUserString(),
+			},
+		},
+		Bin: binary,
+	}
+}
+
+func NewThorchain(testName string, chainConfig ibc.ChainConfig, numValidators int, numFullNodes int, log *zap.Logger) *Thorchain {
+	// if numValidators != 1 {
+	// 	panic(fmt.Sprintf("Thorchain must start with 1 validators for vault and router contract setup"))
```

Comment:
Is this still the case?

Reply:
Yes, when starting a chain from genesis and we want bifrost support for outgoing txs on exogeneous chains (i.e. for native swaps, withdraws, ejecting savers, or ragnarok events / all functionality). But no, when we are starting from an exported state (i.e. mainnet) in order to test genesis restart hard forks (not testing much more than getting the chain to produce blocks). The code was commented out for the hard fork test. Since the vaults are controlled by TSS keys that we don't own, we'll need additional logic to have the nodes create new vault addresses by the validators starting up the hard fork and in theory, we could apply that to a normal genesis start too in order to support more than one validator.